### PR TITLE
[DFSM] Support live updates on compute nodes.

### PR DESCRIPTION
**Pull request by gmarciani**
### Description of changes
Add support for live updates on compute nodes, using cfn-hup. In particular:
* Move definition of dna.json from compute user data to compute launch template. In this way changes to cluster config file, which are reflected to the dna.json, are reflected on the launch template and so detectable by cfn-hup.
* Add cfn-hup configuration files.
* Add cluster config versioning to the dna.json.
* Add permissions for compute nodes to describe the resources in the compute fleet nested stack, required by cfn-init and cfn-hup to describe the stack and detect updates.
* Add permissions for compute nodes to read/write items in the DDB table, required to put the version of the cluster config file deployed in the nodes.
* Relaxed the unit test `test_cluster_config_limits` to unblock these changes by reducing the maximum number of queues validated by the test from 50 to 46. We will revert this relaxation in a follow up PR where we will be able to move the content causing the limit breach from the template to the AMI.
* Added new unit tests covering IAM permissions for compute nodes and generation of dna.json (generation of cfn-hup config file intentionally not covered by unit tests because we should move it to cookbook in the short term).
* Ignore bandit rules B036 and B038, that started failing in this PR but not due to the changes included in this PR (all B038 findings are also false positive)

This PR relies on https://github.com/aws/aws-parallelcluster-cookbook/pull/2614

### Tests
* Existing unit tests (some of them are still failing, this is the reason why the Pr is still in draft)
* New unit tests covering IAM permissions for compute nodes and generation of dna.json.
* Manual testing: cluster creation + forced update verifying that the update succeeds and the log line proving it is emitted see https://github.com/aws/aws-parallelcluster-cookbook/pull/2614).
* Integ test test `test_update_slurm` succeeded.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/resources/compute_node/user_data.sh [hunk 1]
```diff
@@ -166,6 +130,10 @@ write_files:
         fi
       }
 
+      export PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/opt/aws/bin
+
+      cfn-init -s ${AWS::StackName} -v -c deployFiles -r ${LaunchTemplateResourceId} --region ${AWS::Region} --url ${CloudFormationUrl} --role ${CfnInitRole} || error_exit 'Failed to bootstrap the compute node. Please check /var/log/cfn-init.log or /var/log/chef-client.log in the compute node, or check the cfn-init.log or chef-client.log in CloudWatch logs. Please refer to https://docs.aws.amazon.com/parallelcluster/latest/ug/troubleshooting-v3.html#troubleshooting-v3-get-logs for more details on ParallelCluster logs.'
```
Comment (enrico-usai) [inline cli/src/pcluster/resources/compute_node/user_data.sh]: chef-client.log is not available in the compute node.

File: cli/tests/pcluster/templates/test_queues_stack.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+from tests.pcluster.aws.dummy_aws_api import mock_aws_api
+
+from tests.pcluster.models.dummy_s3_bucket import mock_bucket_object_utils, dummy_cluster_bucket
+
+from pcluster.schemas.cluster_schema import ClusterSchema
+from pcluster.templates.cdk_builder import CDKTemplateBuilder
+from pcluster.utils import load_yaml_dict, load_json_dict
+from tests.pcluster.templates.test_cluster_stack import get_generated_template_and_cdk_assets, IamPolicyAssertion, \
+    _get_cfn_init_file_content
```
[the remark this answers is not in the document]
Comment (gmarciani) [reply cli/tests/pcluster/templates/test_queues_stack.py]: Fixed

File: cli/tests/pcluster/templates/test_queues_stack.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+from tests.pcluster.utils import get_asset_content_with_resource_name, get_resource_from_assets, render_user_data, \
+    validate_dna_json_fields
```
[the remark this answers is not in the document]
Comment (gmarciani) [reply cli/tests/pcluster/templates/test_queues_stack.py]: Fixed

File: cli/src/pcluster/resources/compute_node/user_data.sh (hunk 1 above)
Comment (gmarciani) [reply cli/src/pcluster/resources/compute_node/user_data.sh]: Thanks for the heads up. I'll fix the log line.

File: tests/integration-tests/tests/storage/snapshots_factory.py [hunk 4]
```diff
@@ -157,7 +157,7 @@
                 )
                 ssh_conn.open()
                 tries = 0
-            except BaseException:
+            except BaseException:  # noqa: B036
```
[the remark this answers is not in the document]
Comment (gmarciani) [reply tests/integration-tests/tests/storage/snapshots_factory.py]: The except statement was not introduced by this PR.

[the remark this answers is not in the document]
Comment (enrico-usai) [reply tests/integration-tests/tests/storage/snapshots_factory.py]: I'd say we can simply use "Exception" here.

File: cli/tests/pcluster/templates/test_queues_stack.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+@freeze_time("2024-01-15T15:30:45")
+@pytest.mark.parametrize(
+    "config_file_name, expected_compute_node_dna_json_file_name, expected_compute_node_extra_json_file_name",
+    [
+        ("config-1.yaml", "dna-1.json", "extra-1.json"),
+        ("config-2.yaml", "dna-2.json", "extra-2.json"),
+    ],
+)
+def test_compute_nodes_dna_json(
```
Comment (enrico-usai) [inline cli/tests/pcluster/templates/test_queues_stack.py:125]: thanks for creating this unit test!

**Review by enrico-usai — APPROVED** [review]
The change looks good to me. I verified that all the variables used in the user-data.sh of the compute node are still populated and that we just moved the variables required for dna.json.

**Pull request merged**